File: tests/fixtures/benchmark/samples/excellent_binary_search.py
def binary_search(arr, target):
    """Find target in sorted array using binary search.

    Returns index if found, -1 otherwise.
    Time: O(log n). Space: O(1).
    """
    low, high = 0, len(arr) - 1
    while low <= high:
        mid = (low + high) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            low = mid + 1
        else:
            high = mid - 1
    return -1

File: tests/fixtures/benchmark/samples/excellent_binary_search.go
package benchmark

import "cmp"

func BinarySearch[T cmp.Ordered](items []T, target T) int {
	low, high := 0, len(items)-1
	for low <= high {
		mid := low + (high-low)/2
		switch {
		case items[mid] == target:
			return mid
		case items[mid] < target:
			low = mid + 1
		default:
			high = mid - 1
		}
	}
	return -1
}

func BinarySearchFunc[T any](items []T, target T, compare func(a, b T) int) int {
	low, high := 0, len(items)-1
	for low <= high {
		mid := low + (high-low)/2
		switch c := compare(items[mid], target); {
		case c == 0:
			return mid
		case c < 0:
			low = mid + 1
		default:
			high = mid - 1
		}
	}
	return -1
}
